Update readme setup

groovecoder opened the pull request.

groovecoder commented:
@rtilder you good to merge this?

groovecoder commented:
@ckolos is this a crazy-dumb change to `Dockerfile` and `shavar.ini`?

1. I added a `COPY requirements*.txt` above the `RUN pip install` so that docker will cache installed requirements and can skip installing requirements unless/until the `requires*.txt` files change. It helps speed up builds in dev especially.

2. I also updated the `README.md` instructions to `docker build` and `docker run`

3. I changed `shavar.ini` `[uwsgi] http` to bind to `0.0.0.0` so I can curl `localhost` from host machine

ckolos commented:
I think the uwsgi change is going to break the circleci test on the container build... passing -host=net on the command might work as well and wouldn't disrupt circle. I think long-term we should make shavar/uwsgi look at the ENV vars passed to make the decision on which interface to use

groovecoder commented:
Ironically, Circle CI is working while Travis CI was failing. I've fixed them both, but I'm going to try `-host=net` on the command anyway because that feels like the right solution.

groovecoder commented:
Or rather, a better solution.

bytesized commented:
Running tests on my machine from this branch (including using the test instructions in this branch's README) works for me. I do not know what to do with the instruction to "make sure your ~/.boto config works" as I have never heard of boto, but it seemed to work fine nevertheless.

groovecoder commented:
@ckolos - does this look good again? I reverted the `shavar.ini` changes and Circle, Travis, and local docker dev environments are all working.

bytesized commented:
I just tried using the Docker run instructions on this branch, which did not seem to work:

```
$ curl -v -d" " localhost:8080/list
* Hostname was NOT found in DNS cache
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 8080 (#0)
> POST /list HTTP/1.1
> User-Agent: curl/7.35.0
> Host: localhost:8080
> Accept: */*
> Content-Length: 1
> Content-Type: application/x-www-form-urlencoded
> 
* upload completely sent off: 1 out of 1 bytes
* Recv failure: Connection reset by peer
* Closing connection 0
curl: (56) Recv failure: Connection reset by peer
```

Docker output:

```
[uWSGI] getting INI configuration from shavar.ini
*** Starting uWSGI 2.0.11.1 (64bit) on [Fri Nov 11 17:34:18 2016] ***
compiled with version: 4.9.2 on 11 November 2016 17:33:10
os: Linux-4.2.0-27-generic #32~14.04.1-Ubuntu SMP Fri Jan 22 15:32:26 UTC 2016
nodename: 37f799b779bb
machine: x86_64
clock source: unix
pcre jit disabled
detected number of CPU cores: 8
current working directory: /app
detected binary path: /usr/local/bin/uwsgi
uWSGI running as root, you can use --uid/--gid/--chroot options
setgid() to 10001
setuid() to 10001
chdir() to /app
*** WARNING: you are running uWSGI without its master process manager ***
your processes number limit is 524288
your memory page size is 4096 bytes
detected max file descriptor number: 524288
lock engine: pthread robust mutexes
thunder lock: disabled (you can enable it with --thunder-lock)
uWSGI http bound on 127.0.0.1:8080 fd 4
spawned uWSGI http 1 (pid: 19)
uwsgi socket 0 bound to TCP address 127.0.0.1:36781 (port auto-assigned) fd 3
Python version: 2.7.12 (default, Nov 10 2016, 08:18:45)  [GCC 4.9.2]
Python main interpreter initialized at 0x2053000
python threads support enabled
your server socket listen backlog is limited to 100 connections
your mercy for graceful operations on workers is 60 seconds
mapped 332288 bytes (324 KB) for 8 cores
*** Operational MODE: preforking+threaded ***
added local/bin to pythonpath.
Loading paste environment: config:/app/shavar.ini
WSGI app 0 (mountpoint='') ready in 0 seconds on interpreter 0x2053000 pid: 16 (default app)
*** uWSGI is running in multiple interpreter mode ***
spawned uWSGI worker 1 (pid: 16, cores: 2)
spawned uWSGI worker 2 (pid: 24, cores: 2)
spawned uWSGI worker 3 (pid: 25, cores: 2)
spawned uWSGI worker 4 (pid: 27, cores: 2)
```

I am running this on an Ubuntu VM:
```
$ uname -a
Linux ubuntu 4.2.0-27-generic #32~14.04.1-Ubuntu SMP Fri Jan 22 15:32:26 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
```

groovecoder commented:
Did you do `docker run -p 8080:8080 <image-hash>` ?

bytesized commented:
Yes. That resulted in the "Docker Output" in my last comment. The output did not change when I ran `curl -v -d" " localhost:8080/list` in another terminal.

bytesized commented:
I have gotten the existing master branch working with:

```
mkdir bin
ln -s `which flake8` bin/flake8
ln -s `which nosetests` bin/nosetests
make test
python setup.py develop
pserve shavar.testing.ini
```

```
$ curl -v -d" " localhost:6543/list
* Hostname was NOT found in DNS cache
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 6543 (#0)
> POST /list HTTP/1.1
> User-Agent: curl/7.35.0
> Host: localhost:6543
> Accept: */*
> Content-Length: 1
> Content-Type: application/x-www-form-urlencoded
> 
* upload completely sent off: 1 out of 1 bytes
< HTTP/1.1 200 OK
< Content-Length: 38
< Content-Type: text/plain; charset=UTF-8
< Date: Fri, 11 Nov 2016 18:03:02 GMT
* Server waitress is not blacklisted
< Server: waitress
< 
moz-abp-shavar
mozpub-track-digest256
* Connection #0 to host localhost left intact
```

groovecoder commented:
With the docker container running, what do your NAT tables look like? (https://docs.docker.com/engine/userguide/networking/default_network/binding/)

bytesized commented:
```
$ sudo iptables -t nat -L -n
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination         
DOCKER     all  --  0.0.0.0/0            0.0.0.0/0            ADDRTYPE match dst-type LOCAL

Chain INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         
DOCKER     all  --  0.0.0.0/0           !127.0.0.0/8          ADDRTYPE match dst-type LOCAL

Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination         
MASQUERADE  all  --  172.17.0.0/16        0.0.0.0/0           
MASQUERADE  all  --  172.18.0.0/16        0.0.0.0/0           
MASQUERADE  tcp  --  172.17.0.2           172.17.0.2           tcp dpt:8080
MASQUERADE  tcp  --  172.17.0.2           172.17.0.2           tcp dpt:8080

Chain DOCKER (2 references)
target     prot opt source               destination         
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:8080 to:172.17.0.2:8080
```

groovecoder commented:
Can you try to `docker build .` and `docker run -p 8080:8080 <image>` again? I changed `server:main` and `uwsgi` configs to port 8080 now.

bytesized commented:
Ah! That seems to have done the trick.

groovecoder commented:
@ckolos - I changed the `uwsgi` config again, but this time I just changed it to bind to `0.0.0.0`; i.e., I didn't change the port.

It passes CircleCI, so I think we're good here now?

ckolos commented:
no issues. r+
